ietf/doc/migrations/0024_remove_ad_is_watching_states.py

```diff
@@ -0,0 +1,121 @@
+            return dict((field.name, getattr(obj, field.name))
+                        for field in obj._meta.fields
+                        if field is not obj._meta.pk)
+    
+        # copy fields
+        fields = get_model_fields_as_dict(doc)
+        fields["doc"] = doc
+        fields["name"] = doc.name
+    
+        dochist = DocHistory(**fields)
+        dochist.save()
+    
+        # copy many to many
+        for field in doc._meta.many_to_many:
+            if field.remote_field.through and field.remote_field.through._meta.auto_created:
+                hist_field = getattr(dochist, field.name)
+                hist_field.clear()
+                hist_field.set(getattr(doc, field.name).all())
+    
+        # copy remaining tricky many to many
+        def transfer_fields(obj, HistModel):
+            mfields = get_model_fields_as_dict(item)
+            # map doc -> dochist
+            for k, v in mfields.items():
+                if v == doc:
+                    mfields[k] = dochist
+            HistModel.objects.create(**mfields)
+    
+        for item in RelatedDocument.objects.filter(source=doc):
+            transfer_fields(item, RelatedDocHistory)
+    
+        for item in DocumentAuthor.objects.filter(document=doc):
+            transfer_fields(item, DocHistoryAuthor)
+    
+        return dochist
+    
+    return save_document_in_history
+
+
+def forward(apps, schema_editor):
+    """Mark watching draft-iesg state unused after removing it from Documents"""
+    StateDocEvent = apps.get_model("doc", "StateDocEvent")
+    Document = apps.get_model("doc", "Document")
+    State = apps.get_model("doc", "State")
+    StateType = apps.get_model("doc", "StateType")
+    Person = apps.get_model("person", "Person")
+    
+    save_document_in_history = get_helper(
+        DocHistory=apps.get_model("doc", "DocHistory"),
+        RelatedDocument=apps.get_model("doc", "RelatedDocument"),
+        RelatedDocHistory=apps.get_model("doc", "RelatedDocHistory"),
+        DocumentAuthor=apps.get_model("doc", "DocumentAuthor"),
+        DocHistoryAuthor=apps.get_model("doc", "DocHistoryAuthor"),
+    )
+
+    draft_iesg_state_type = StateType.objects.get(slug="draft-iesg")
+    idexists_state = State.objects.get(type=draft_iesg_state_type, slug="idexists")
+    watching_state = State.objects.get(type=draft_iesg_state_type, slug="watching")
+    system_person = Person.objects.get(name="(System)")
+
+    # Remove state from documents that currently have it
+    for doc in Document.objects.filter(states=watching_state):
+        assert doc.type_id == "draft"
+        doc.states.remove(watching_state)
+        doc.states.add(idexists_state)
+        e = StateDocEvent.objects.create(
+            type="changed_state",
+            by=system_person,
+            doc=doc,
+            rev=doc.rev,
+            desc=f"{draft_iesg_state_type.label} changed to <b>{idexists_state.name}</b> from {watching_state.name}",
```

Comment:
cringing at the html tags, but keeping the set of things to mutate later regular was a good call